boards/rp-pico/examples/pico_pio_pwm.rs

```diff
@@ -0,0 +1,162 @@
+//! # Pico PIO PWM Blink Example
+//!
+//! Fades the LED on a Pico board using the PIO peripheral with an pwm program.
+//!
+//! This will fade in the LED attached to GP25, which is the pin the Pico
+//! uses for the on-board LED.
+//!
+//! This example uses a few advance pio tricks such as side setting pins and instruction injection.
+//!
+//! See the `Cargo.toml` file for Copyright and license details. Except for the pio program which is subject to a different license.
+
+#![no_std]
+#![no_main]
+
+use defmt::info;
+use defmt_rtt as _;
+// The macro for our start-up function
+use rp_pico::entry;
+
+// Time handling traits
+use embedded_time::rate::*;
+
+// Ensure we halt the program on panic (if we don't mention this crate it won't
+// be linked)
+use panic_halt as _;
+
+// Pull in any important traits
+use rp_pico::hal::prelude::*;
+
+// A shorter alias for the Peripheral Access Crate, which provides low-level
+// register access
+use rp_pico::hal::pac;
+
+// A shorter alias for the Hardware Abstraction Layer, which provides
+// higher-level drivers.
+use rp_pico::hal;
+
+// Import pio crates
+use hal::pio::{PIOBuilder, Running, StateMachine, Tx, ValidStateMachine, SM0};
+use pio::{InstructionOperands, OutDestination};
+use pio_proc::pio_file;
+
+/// Set pio pwm period
+///
+/// This uses a sneaky trick to set a second value besides the duty cycle.
+/// We first write a value to the tx fifo. But instead of the normal instructions we
+/// have stopped the state machine and inject our own instructions that move the written value to the ISR.
+fn pio_pwm_set_period<T: ValidStateMachine>(
+    sm: StateMachine<(hal::pac::PIO0, SM0), Running>,
+    tx: &mut Tx<T>,
+    period: u32,
+) -> StateMachine<(hal::pac::PIO0, SM0), Running> {
+    let mut sm = sm.stop();
+    tx.write(period);
```

Comment:
Calling `pio_pwm_set_period` shortly after `pio_pwm_set_level` could have unintended results: If the value written to the tx fifo by `pio_pwm_set_level` has not yet been read by the PIO program before the state machine is stopped, the level will be interpreted as a period.

Perhaps add `while !tx.is_empty() {}` before the call to `sm.stop()`?

Reply:
Interesting.. You are completely right. I based the example on https://github.com/raspberrypi/pico-examples/blob/master/pio/pwm/pwm.c which has the same issue...